File: 2744-find-maximum-number-of-string-pairs/2744-find-maximum-number-of-string-pairs.ts
function reverseString(s: string): string {
  let result = "";
  for (let i = s.length - 1; i >= 0; i--) {
    result += s[i];
  }
  return result;
}

function maximumNumberOfStringPairs(words: string[]): number {
  const memoi = new Set<string>();
  let result = 0;

  words.forEach((s) => {
    const reversed = reverseString(s);
    if (memoi.has(reversed)) {
      result++;
    }
    memoi.add(s);
  });

  return result;
}

File: 2744-find-maximum-number-of-string-pairs/2744-find-maximum-number-of-string-pairs.go
func reverseString(s string) string {
	var result strings.Builder

	for i := utf8.RuneCountInString(s) - 1; i >= 0; i-- {
		element := s[i]
		result.WriteByte(element)
	}

	return result.String()
}

func maximumNumberOfStringPairs(words []string) int {
	usedIndex := make(map[int]bool)
	result := 0

	for i := 0; i < len(words)-1; i++ {
		for j := i + 1; j < len(words); j++ {

			_, isIExist := usedIndex[i]
			_, isJExist := usedIndex[j]

			if isIExist || isJExist {
				continue
			}

			wordI := words[i]
			wordJ := reverseString(words[j])

			if wordI == wordJ {
				result++
				usedIndex[i] = true
				usedIndex[j] = true
			}

		}
	}

	return result
}
